pkg/controller/certificaterequests/acme/acme.go

```diff
@@ -0,0 +1,213 @@
+
+	orderLister cmlisters.OrderLister
+	cmClientV   cmclientset.CertmanagerV1alpha1Interface
+	helper      issuerpkg.Helper
+}
+
+func init() {
+	// create certificate request controller for acme issuer
+	controllerpkg.Register(CRControllerName, func(ctx *controllerpkg.Context) (controllerpkg.Interface, error) {
+		acme := NewACME(ctx)
+
+		orderInformer := ctx.SharedInformerFactory.Certmanager().V1alpha1().Orders().Informer()
+		controller := certificaterequests.New(apiutil.IssuerACME, acme, orderInformer)
+
+		c, err := controllerpkg.New(ctx, CRControllerName, controller)
+		if err != nil {
+			return nil, err
+		}
+
+		return c.Run, nil
+	})
+}
+
+func NewACME(ctx *controllerpkg.Context) *ACME {
+	ctx.CMClient.CertmanagerV1alpha1()
+	return &ACME{
+		recorder:      ctx.Recorder,
+		issuerOptions: ctx.IssuerOptions,
+		orderLister:   ctx.SharedInformerFactory.Certmanager().V1alpha1().Orders().Lister(),
+		cmClientV:     ctx.CMClient.CertmanagerV1alpha1(),
+	}
+}
+
+func (a *ACME) Sign(ctx context.Context, cr *v1alpha1.CertificateRequest, issuer v1alpha1.GenericIssuer) (*issuerpkg.IssueResponse, error) {
+	log := logf.FromContext(ctx, "sign")
+	reporter := crutil.NewReporter(cr, a.recorder)
+	resourceNamespace := a.issuerOptions.ResourceNamespace(issuer)
+
+	// If we can't decode the CSR PEM we have to hard fail
+	csr, err := pki.DecodeX509CertificateRequestBytes(cr.Spec.CSRPEM)
+	if err != nil {
+		message := "Failed to decode CSR in spec"
+
+		reporter.Failed(err, "ErrorParsingCSR", message)
+		log.Error(err, message)
+
+		return nil, nil
+	}
+
+	// If we fail to build the order we have to hard fail.
+	expectedOrder, err := buildOrder(cr, csr)
+	if err != nil {
+		message := "Failed to build order"
+
+		reporter.Failed(err, "ErrorBuildingOrder", message)
+		log.Error(err, message)
+
+		return nil, nil
+	}
+
+	order, err := a.orderLister.Orders(expectedOrder.Namespace).Get(expectedOrder.Name)
+	if err != nil {
+		if k8sErrors.IsNotFound(err) {
+			// Failing to create the order here is most likely network related.
+			// We should backoff and keep trying.
+			order, err = a.cmClientV.Orders(resourceNamespace).Create(expectedOrder)
+			if err != nil {
+				message := fmt.Sprintf("Failed create new order resource %s/%s", resourceNamespace, expectedOrder.Name)
+
+				reporter.Pending(err, "ErrorCreatingOrder", message)
+				log.Error(err, message)
+
+				return nil, err
+			}
+
+		} else {
+			// We are probably in a network error here so we should backoff and retry
+			message := fmt.Sprintf("Failed to get order resource %s/%s", resourceNamespace, expectedOrder.Name)
+
+			reporter.Pending(err, "ErrorGettingOrder", message)
+			log.Error(err, message)
+
+			return nil, err
+		}
+	}
+
+	// If the acme order has failed then so too does the CertificateRequest meet the same fate.
+	if acme.IsFailureState(order.Status.State) {
+		message := fmt.Sprintf("Failed to resolve order resource %s/%s", resourceNamespace, expectedOrder.Name)
```

Comment:
Is the wording of this message similar to the existing wording? It's confusing if we change event messages, and I don't think we use the term 'resolve' in any other events at the moment